### DISCUSSION TITLE
Update process

### DISCUSSION THREAD

**Discussion by leuchtdiode** [pimcore/pimcore#13071] [Q&A / Support]
Hi,

I wondered how you guys are describing the "ideal update process". Because sometimes when composer is updating dependencies and the cache is not yet cleared, the page throws HTTP response 500.

Are you enabling maintenance mode during an update? Because it is no where stated in the docs to do so.

Thx & BR,
Alex

**Comment by BlackbitDevs** [comment]
Imho, you should never run `composer update` on the live system. 
Instead set up an equal environment as the live system (e.g. by cloning live system or having a docker image which you can easily run somewhere else). Do the Pimcore update and all necessary steps on your application code on this cloned system. You will end up with an updated composer.lock file and application code which is compatible with updated Pimcore version. You then put these changed files to the live server, execute `composer install` to install the libraries from composer.lock, run `bin/console doctrine:migrations:migrate` (or [put this into your composer.json to be run automatically](https://pimcore.com/docs/pimcore/current/Development_Documentation/Development_Tools_and_Details/Migrations.html#page_Run-all-available-migrations-after) - but for `post-install-cmd`) and perhaps also `pimcore:deployment:classes-rebuild` which updates database structure according to data object class definition files. This way downtime will be very low. Pessimistic people would set up another clone of the live system and assure that the described deployment process really works or if something has been forgotten. 
This whole cloning process has the advantage that you can take as long as you need to update the system without the live system being down. And if you are an agency you can ask your client to approve the changes.

One side note: You could be saying that the deployment process could be skipped if we use the cloned system as the live system after update. The problem with this approach is that all the data which got changed in the meantime after the clone was created and "now" would get lost.

**Comment by leuchtdiode** [comment]
Alright, thank you for clarification. Sometimes the DataObject classes get adapted after `composer update`. If I committ and push them, could it be that the LIVE system is not working any longer? I'm just curiours if there are any caveats one have to be prepared for.

And maybe the doc could be updated with your suggestion? I think for newcomers the update process is not as clear as it was when the was still the "online update" within pimcore (pre composer). It was also documented well these days when I remember correctly.

Thx & BR,
Alex

**Comment by BlackbitDevs** [comment]
> Sometimes the DataObject classes get adapted after `composer update`

This can only happen when a migration resaves the classes.
Manual changes to the data object class definitions do not get deployed during `composer update` but only with `pimcore:deployment:classes-rebuild`.

The above description of course depends on the infrastructure. If someone uses container-based hosting, the process will be a little different. But let's see if I can create a general approach which describes a possible deployment strategy for all the different hosting setups...

**Reply by leuchtdiode** [reply]
Alright, thank you.
